README.rst
==========
AryaLogger
==========

Use the SimpleAciUiLogServer and arya to convert UI REST API calls to ACI
Python SDK code.

Install
-------

Installation of AryaLogger depends on several other modules that are not part
of the Python standard library.  Here is a basic list of those modules:

1. acicobra - Not yet available on Pypi
2. arya - available on Pypi
3. SimpleAciUiLogServer - available on Pypi

Installing Dependancies
+++++++++++++++++++++++

The difficult dependancy at this point is acicobra as it is not on Pypi yet.

Installing acicobra
"""""""""""""""""""

If you have at least APIC version 1.0(3f) installed on your APIC, installation
of acicobra can be done by simply pointing pip at your APIC's cobra/_downloads
directory using the --find-links option using one of these two commands:

    easy_install -Z --find-links http://apic/cobra/_downloads acicobra

Or:

    easy_install -Z --find-links https://apic/cobra/_downloads acicobra

If your APIC has http disabled (the default), you must either use https or
enable http on the APIC.  Note:  As of the 1.1(1j) version of the APIC,
TLSv1 has been disabled by default on the APIC for HTTPS.  So if you are
using HTTPS and your python install uses an older openssl that does not
support TLSv1_1 or TLSv1_2 you may have issues handshaking with the APIC
GUI.  This is especially the case on the default installation of python on
MacOSX which usually utilizes the default installation of openssl which is
extremely old.  Generally speaking, one should not be using such an old
version of openssl.  Homebrew may be your best bet.

If you have an older version of software on your APIC, you can either download
acicobra from the Cisco.com software download site (APIC specific contract
required to be tied to the Cisco.com id doing the download) or you can download
the acicobra egg from the APIC itself and rename it:

    https://developer.cisco.com/media/apicDcPythonAPI_v0.1/install.html

Installing arya
"""""""""""""""

Installing arya from pypi is as simple as using pip:

    pip install arya

Installing AryaLogger
+++++++++++++++++++++

AryaLogger is available on pypi so installation is fairly straight forward:

    pip install AryaLogger

This should also pull in SimpleAciUiLogServer if you don't have it installed
already.

Run AryaLogger
--------------

There are two possible commands to run AryaLogger on your system.  You only need
to run one of these commands:

    aryalogger

Or:

    AryaLogger

There are additional options available that can be seen via the -h option:

    * -h: Help
    * -a/--apicip: The ip address or hostname of an APIC to help resolve which
      ip address to print when showing the server is started.
    * -po/--port: The port for the HTTP server to listen on - default is 8987
    * -l/--location: The URI path to listen for - default is apiinspector
    * -s/--sslport: The port for the HTTPS server to listen on - default is
      8443
    * -c/--cert: The certificate file to use for TLS/SSL negotiation.
    * -e/--exclude: There are three types of queries the GUI doe constantly
      to keep the session up or to request information to update the GUI.  This
      option allows you to selectively exlude these queries from the output.
      options are: 

      - subscriptionRefresh - Refreshes a subscription request for the GUI.
      - aaaRefresh - Refreshes the GUI connection to the APIC.
      - topInfo - Retrieves basic info from the APIC to update the GUI.

    * -r/--logrequests: Log the HTTP request along with the CobraSDK code.
    * -n/--nice-output: Pretty print any XML or JSON (has no affect currently
      in AryaLogger)
    * -i/--indent: The number of spaces to indent when pretty printing (has
      no affect currently in AryaLogger)

For HTTPS, AryaLogger does come with a default certificate file but it should
not be used for production.  Instead you should create your own certificate file
using openssl and pass it to the server:

    openssl req -new -x509 -keyout server.pem -out server.pem -days 36500 -nodes

Pass it in to AryaLogger using the -c/--cert option.

Setup your APIC
---------------

Setting up your APIC to send log messages to AryaLogger is very simple as well.
When you run AryaLogger it will output the addresses and ports it is listening
to:

    $ aryalogger
    serving at:
    http://10.1.1.100:8987/apiinspector
    https://10.1.1.100:8443/apiinspector

If you are going to be using the https address to access the AryaLogger server,
be sure to browse to the https address printed out by the server when it starts
so that you can accept the certificate exception.

If you connect to the APIC graphical user interface (GUI) via http, take note of
the http URL otherwise if you connect to the GUI using https take note of
the https URL.

Use this URL in the 'Start Remote Logging' pop-up that is available from the 
'Welcome' menu at the top right of the GUI.

Once you have started remote logging to the URL provided by AryaLogger, when you
click around the APIC GUI you should see auto generated Cobra SDK code from
AryaLogger.

Problems & Issues
-----------------

The most common problem seen when running AryaLogger is that the AryaLogger
server starts, remote logging is setup on the APIC but then no data is seen
ever being transferred to the server.  This can be caused by many things but
the most common is that a security exception has not been accepted for the http
certificate.  This can be resolved by connecting to the AryaLogger https
address from your browser and accepting the security exception for the server
certificate.  The second most common issue is that there is a firewall running
on the host that the AryaLogger server is running on.  We have even seen hosts
with multiple firewalls running when the end user had no idea.

The simplest form of troubleshooting involves going into the same browser
tab that the APIC is connected to with and opening the developer tools in
your browser.  Look at the javascript console and see if any errors are being
printed about communication to the host the AryaLogger is running on.

If you run into issues please feel free to open an issue on github.


CHANGES.txt
Changes
=======

v1.1.7
------

* Fixes
  - Added version requirements for SimpleAciUiLogServer and arya to setup.py.
* Enhancements
  - Generalize some of the logging functions so they can be reused between
    methods
  - Add aaaLogout handler
  - Added support for starting a single server (either http or https) via the
    -si/--single-server {http,https} CLI option.
  - Added support for request type filtering via the -ty/--request-type option
  - Added additional exclude options: HDfabricOverallHealth5min-0, aaaLogout and
    all.  all will exclude all potentially noisy queries.
  - Bumped version to 1.1.7

v1.1.6
------

* Fixes
  - #5 Set the default sslport from argument parsing, without this AryaLogger
    crashes when executed from the CLI
* Enhancements
  - Add CHANGES.txt

v1.1.5
------

* Fixes
  - #4 Fixes cobra query strings to be printed properly
  - Address many pylint and landscape.io identified issues
  - Fix an issue with naming_vals
* Enhancements
  - #4 Adds support for the REST API options page, page-size, order-by and replica

v1.1.4
------

* Fixes
  - Simple version bump
* Enhancements
  - None

v1.1.3
------

* Fixes
  - #2 changed query options logic to be less strict
* Enhancements
  - Add additional info to the README about problems & and issues
  - Add info about using self-signed cert and what to do after starting the
    server

v1.1.2
------

* Fixes
  - None
* Enhancements
  - Add a Manifest.in so the LICENSE.txt file is available during install

v1.1.1
------

* Fixes
  - Fix README.rst for now, update setup.py and bump the minor version
* Enhancements
  - None

v1.1.0
------

* Fixes
  - Update setup.py to point to the correct repo
* Enhancements
  - Update ArayLogger to the latest features of SimpleAciUiLogServer
  - Add query string parsing
  - Pep8 changes and correcting direct access to rn related info

v1.0.0
------

Original commit to github

* Fixes
  - None
* Enhancements
  - None

AryaLogger/convert_apic_dn_to_cobra.py
#!/usr/bin/env python
"""
An example of converting from a dn string to APIC Cobra Python SDK.

Written by Mike Timm (mtimm@cisco.com)

Copyright (C) 2014 Cisco Systems Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from urlparse import urlparse, ResultMixin
from collections import OrderedDict, namedtuple
from cobra.mit.naming import Dn


class ApicParseResult(namedtuple('ApicParseResult',
                                 'scheme netloc path params query fragment'),
                      ResultMixin):

    """ApicParseResult.

    Mixin type of class that adds some apic specific properties to the urlparse
    named tuple
    """

    @property
    def dn_or_class(self):
        """Return the dn or the class."""
        pathparts = self._get_path_parts()
        if pathparts[1] != 'node':
            return self._get_dn_or_class(pathparts, 1)
        else:
            return self._get_dn_or_class(pathparts, 2)

    @property
    def api_format(self):
        """Return the api format."""
        return self._get_api_format(self.path)

    @property
    def api_method(self):
        """Return the api method."""
        pathparts = self._get_path_parts()
        if pathparts[1] == 'node':
            return pathparts[2]
        else:
            return pathparts[1]

    @property
    def classnode(self):
        """Return the class node."""
        if self.api_method != 'class':
            return ""
        pathparts = self._get_path_parts()
        if pathparts[1] != 'node':
            return self._get_classnode(pathparts, 3)
        else:
            return self._get_classnode(pathparts, 4)

    @staticmethod
    def _get_classnode(parts, index):
        """Get the class node."""
        if len(parts) <= index:
            return ""
        else:
            return "/".join(parts[index - 1:-1])

    def _get_path_parts(self):
        """Break the path up into a list and return it."""
        dn_str = self._remove_format_from_path(self.path, self.api_format)
        return dn_str[1:].split("/")

    @staticmethod
    def _remove_format_from_path(path, fmt):
        """Remove the api format from the path."""
        return path[:-len("." + fmt)]

    @staticmethod
    def _get_api_format(path):
        """Get the api format."""
        if path.endswith(".xml"):
            return 'xml'
        elif path.endswith(".json"):
            return 'json'

    @staticmethod
    def _get_dn_or_class(parts, index):
        """Get the dn or the class depending on the type of query."""
        if parts[index] == 'class':
            return parts[-1]
        elif parts[index] == 'mo':
            return "/".join(parts[index + 1:])
        else:
            return ""


def apic_rest_urlparse(url_str):
    """Parse an ACI REST API URL."""
    tpl = urlparse(url_str)
    scheme, netloc, path, params, query, fragment = tpl
    return ApicParseResult(scheme, netloc, path, params, query, fragment)


def convert_dn_to_cobra(dn_str):
    """Convert an ACI distinguished name to ACI Python SDK code."""
    cobra_dn = Dn.fromString(dn_str)
    parent_mo_or_dn = "''"
    dn_dict = OrderedDict()
    for rn_obj in cobra_dn.rns:
        rn_str = str(rn_obj)
        dn_dict[rn_str] = {}
        dn_dict[rn_str]['namingVals'] = list(rn_obj.namingVals)
        dn_dict[rn_str]['moClassName'] = rn_obj.meta.moClassName
        dn_dict[rn_str]['className'] = rn_obj.meta.className
        dn_dict[rn_str]['parentMoOrDn'] = parent_mo_or_dn
        parent_mo_or_dn = rn_obj.meta.moClassName
    for arn in dn_dict.items():
        if len(list(arn[1]['namingVals'])) > 0:
            nvals = [str(val) for val in arn[1]['namingVals']]
            nvals_str = ", '" + ", ".join(nvals) + "'"
        else:
            nvals_str = ""
        print "{0} = {1}({2}{3})".format(arn[1]['moClassName'],
                                         arn[1]['className'],
                                         arn[1]['parentMoOrDn'],
                                         nvals_str)


if __name__ == '__main__':
    convert_dn_to_cobra('topology/HDfabricOverallHealth5min-0')
    print("")
    convert_dn_to_cobra('uni/tn-mgmt/mgmtp-default/oob-default')
    print("")

    URL = 'https://10.122.254.211/api/node/mo/topology/HDfabricOverallHealt' + \
          'h5min-0.json'

    convert_dn_to_cobra(apic_rest_urlparse(URL).dn_or_class)
    print("")

    URL = 'https://10.122.254.211/api/node/mo/uni/tn-mgmt/mgmtp-default/oob' + \
          '-default.json'
    convert_dn_to_cobra(apic_rest_urlparse(URL).dn_or_class)
    print


AryaLogger/AryaLogger.py
#!/usr/bin/env python
"""AryaLogger.

A small server based on SimpleAciUiLogServer that can convert APIC GUI logging
messages to ACI Python SDK (cobra) code.

Depends on SimpleAciUiLogServer, arya and acicobra/acimodel python modules.
"""

import logging
import os
import signal
import socket
import sys
import tempfile
from collections import OrderedDict, namedtuple
from urlparse import urlparse, ResultMixin, parse_qs
from argparse import ArgumentParser
from SimpleAciUiLogServer import (SimpleAciUiLogServer,
                                  ThreadingSimpleAciUiLogServer,
                                  serve_forever)
from cobra.mit.naming import Dn
from arya import arya

SERVER_CERT = """
-----BEGIN RSA PRIVATE KEY-----
MIICXAIBAAKBgQC+oA+hYsF3uBIMt7i1ELfUFnyf4/MKM/Ylmy4yBc0/YhqANXYk
so3+gAGkgRlv9ODdsFS7KvjzyaT0kjgA3ahDPyvtroAOWsdFdHJvtS4Ek1WI1Bee
0hNZlTmjQgnjp9ENYl9ImGWghcubJhtse5cJhL9c/hq40do4llZjaaEiCQIDAQAB
AoGAYbd1K7qPCErAXeqT8aVXRo4cZm4YeSN3y4FH5faZZyNoCE7feCJbrZl6vhQ7
sOtrldi9JpD5uyaju4d00+TMSoFqnTuAYLg4CEUAkAq2Hgg1EfQfPpC8IgYdR5qQ
hRu0JArXldch1YLHw8GQGkkZe/cJXiHs/FPjmdUQSsydI50CQQDuEecLrSLjuMpE
i8xjD9cQxSDTHJFDZttVb08fjaKFJk993TsUR2t/eR92OR5m0CFei/RUyYpUaPbk
1s3Eau7XAkEAzPtnMMKoGR3qfLqXzfmgLwQA0UbeV8PbxRCkaCnSYcpn0qJH7UtS
Qjb4X6MPA9bNUnydWFgbPgz4MwKRo0q6HwJAP6DxS6GerZZ6GQ/0NJXLOWQ2fbYo
7QbUoGT7lMdaJJQ0ssMqQyVDifJpgkOJ6JjAEnD9gJvNKPpU4py2qkSaSQJANngr
0Jo5XwtDD0fqJPLLbRLsQLBLTxkdoj0s4v0SCahmdGNpJ5ZXUn8W+xryV3vR7bRt
f1dSTefWYH+zQagO0wJBANlNp79CN7ylgXdrhRVQmBsXHN4G8biUUxMYsfK4Ao/i
Ga3xtkYLv7OmrtY+Gx6w56Jqxyucaka8VBHK0/7JTLE=
-----END RSA PRIVATE KEY-----
-----BEGIN CERTIFICATE-----
MIID+jCCA2OgAwIBAgIJALUh5RwHQhJoMA0GCSqGSIb3DQEBBQUAMIGvMQswCQYD
VQQGEwJVUzELMAkGA1UECBMCQ0ExETAPBgNVBAcTCFNhbiBKb3NlMRUwEwYDVQQK
EwxhcGlpbnNwZWN0b3IxHTAbBgNVBAsTFFNpbXBsZUFjaVVpTG9nU2VydmVyMSow
KAYDVQQDEyFTaW1wbGVBY2lVaUxvZ1NlcnZlci5hcGlpbnNwZWN0b3IxHjAcBgkq
hkiG9w0BCQEWD210aW1tQGNpc2NvLmNvbTAgFw0xNTAxMjMwMDI1NDJaGA8zMDE0
MDUyNjAwMjU0Mlowga8xCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTERMA8GA1UE
BxMIU2FuIEpvc2UxFTATBgNVBAoTDGFwaWluc3BlY3RvcjEdMBsGA1UECxMUU2lt
cGxlQWNpVWlMb2dTZXJ2ZXIxKjAoBgNVBAMTIVNpbXBsZUFjaVVpTG9nU2VydmVy
LmFwaWluc3BlY3RvcjEeMBwGCSqGSIb3DQEJARYPbXRpbW1AY2lzY28uY29tMIGf
MA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC+oA+hYsF3uBIMt7i1ELfUFnyf4/MK
M/Ylmy4yBc0/YhqANXYkso3+gAGkgRlv9ODdsFS7KvjzyaT0kjgA3ahDPyvtroAO
WsdFdHJvtS4Ek1WI1Bee0hNZlTmjQgnjp9ENYl9ImGWghcubJhtse5cJhL9c/hq4
0do4llZjaaEiCQIDAQABo4IBGDCCARQwHQYDVR0OBBYEFN2EqumA49KSEPjLLSni
UtKth4zQMIHkBgNVHSMEgdwwgdmAFN2EqumA49KSEPjLLSniUtKth4zQoYG1pIGy
MIGvMQswCQYDVQQGEwJVUzELMAkGA1UECBMCQ0ExETAPBgNVBAcTCFNhbiBKb3Nl
MRUwEwYDVQQKEwxhcGlpbnNwZWN0b3IxHTAbBgNVBAsTFFNpbXBsZUFjaVVpTG9n
U2VydmVyMSowKAYDVQQDEyFTaW1wbGVBY2lVaUxvZ1NlcnZlci5hcGlpbnNwZWN0
b3IxHjAcBgkqhkiG9w0BCQEWD210aW1tQGNpc2NvLmNvbYIJALUh5RwHQhJoMAwG
A1UdEwQFMAMBAf8wDQYJKoZIhvcNAQEFBQADgYEABPx5cxBNOjWOxZbiRVfpzKac
MKs4tFNtEmilAY7kvNouGaSl1Yw2fCpGXjstOG0+SxPy34YgeQSVOGQI1KXhd7vk
nALqxrKiP2rzpZveBkjq5voRpFw2creEXyt76EKQgwRHYJP60Vu3bYnYNoFHdUwE
TOBaHjC6ZZLRd77dd3s=
-----END CERTIFICATE-----
"""


class ApicParseResult(namedtuple('ApicParseResult',
                                 'scheme netloc path params query fragment'),
                      ResultMixin):

    """ApicParseResult class.

    Mixin type of class that adds some apic specific properties to the urlparse
    named tuple
    """

    @property
    def dn_or_class(self):
        """Check if a path is for a dn or class."""
        pathparts = self._get_path_parts()
        if pathparts[1] != 'node':
            return self._get_dn_or_class(pathparts, 1)
        else:
            return self._get_dn_or_class(pathparts, 2)

    @property
    def api_format(self):
        """return the api format portion of the URL."""
        return self._get_api_format(self.path)

    @property
    def api_method(self):
        """Return the api method."""
        pathparts = self._get_path_parts()
        if pathparts[1] == 'node':
            return pathparts[2]
        else:
            return pathparts[1]

    @property
    def classnode(self):
        """Return the class or an empty string for mo queries."""
        if self.api_method != 'class':
            return ""
        pathparts = self._get_path_parts()
        if pathparts[1] != 'node':
            return self._get_classnode(pathparts, 3)
        else:
            return self._get_classnode(pathparts, 4)

    @staticmethod
    def _get_classnode(parts, index):
        """Get the class portion of a path."""
        if len(parts) <= index:
            return ""
        else:
            return "/".join(parts[index - 1:-1])

    def _get_path_parts(self):
        """Return a list of path parts."""
        dn_str = self._remove_format_from_path(self.path, self.api_format)
        dn_str = self._sanitize_path(dn_str)
        return dn_str[1:].split("/")

    @staticmethod
    def _remove_format_from_path(path, fmt):
        """Remove the api format from the path, including the ."""
        return path[:-len("." + fmt)]

    @staticmethod
    def _get_api_format(path):
        """Return either xml or json for the api format."""
        if path.endswith(".xml"):
            return 'xml'
        elif path.endswith(".json"):
            return 'json'

    @staticmethod
    def _get_dn_or_class(parts, index):
        """Return just the dn or the class."""
        if parts[index] == 'class':
            return parts[-1]
        elif parts[index] == 'mo':
            return "/".join(parts[index + 1:])
        else:
            return ""

    @staticmethod
    def _sanitize_path(path):
        """Left strip any / from the path."""
        return path.lstrip("/")


def apic_rest_urlparse(url):
    """Parse the APIC REST API URL."""
    atuple = urlparse(url)
    scheme, netloc, path, params, query, fragment = atuple
    return ApicParseResult(scheme, netloc, path, params, query, fragment)


def convert_dn_to_cobra(dn_str):
    """Convert an ACI distinguished name to ACI Python SDK (cobra) code."""
    cobra_dn = Dn.fromString(dn_str)
    parent_mo_or_dn = "''"
    dn_dict = OrderedDict()
    for rname in cobra_dn.rns:
        rn_str = str(rname)
        dn_dict[rn_str] = {}
        dn_dict[rn_str]['namingVals'] = tuple(rname.namingVals)
        dn_dict[rn_str]['moClassName'] = rname.meta.moClassName
        dn_dict[rn_str]['className'] = rname.meta.className
        dn_dict[rn_str]['parentMoOrDn'] = parent_mo_or_dn
        parent_mo_or_dn = rname.meta.moClassName
    cobra_str = ""
    for arn in dn_dict.items():
        if len(arn[1]['namingVals']) > 0:
            nvals = [str(val) for val in arn[1]['namingVals']]
            nvals_str = ", '" + ", ".join(nvals) + "'"
        else:
            nvals_str = ""
        cobra_str += "    # {0} = {1}({2}{3})\n".format(arn[1]['moClassName'],
                                                        arn[1]['className'],
                                                        arn[1]['parentMoOrDn'],
                                                        nvals_str)
    return cobra_str


def parse_apic_options_string(options):
    """Parse the REST API options string."""
    dictmap = {
        'rsp-prop-include':     'propInclude',
        'rsp-subtree-filter':   'subtreePropFilter',
        'rsp-subtree-class':    'subtreeClassFilter',
        'rsp-subtree-include':  'subtreeInclude',
        'query-target':         'queryTarget',
        'target-subtree-class': 'classFilter',
        'query-target-filter':  'propFilter',
        'rsp-subtree':          'subtree',
        'replica':              'replica',
        'target-class':         'targetClass',
        'page':                 'page',
        'page-size':            'pageSize',
        'order-by':             'orderBy',
        'replica':              'replica',
    }
    qstring = ''
    if options is None or options == '':
        return qstring
    for opt, value in parse_qs(options).items():
        if opt in dictmap.keys():
            val_str = value[0].replace('"', '\"')
            qstring += '    query.{0} = "{1}"\n'.format(dictmap[opt], val_str)
        else:
            qstring += ('    # Query option "{0}" is not'.format(opt) +
                        ' supported by Cobra SDK\n')
    return qstring


def get_dn_query(dn_str):
    """Get the dn query string."""
    cobra_str = "    query = cobra.mit.request.DnQuery('"
    cobra_str += str(dn_str)
    cobra_str += "')"
    return cobra_str


def get_class_query(kls):
    """Get the class query string."""
    cobra_str = "    query = cobra.mit.request.ClassQuery('"
    cobra_str += str(kls)
    cobra_str += "')"
    return cobra_str


def handle_mo(purl, qstring):
    cobra_str2 = convert_dn_to_cobra(purl.dn_or_class)
    cobra_str2 += "    # Direct dn query:\n"
    cobra_str2 += get_dn_query(purl.dn_or_class)
    cobra_str2 += "\n"
    cobra_str = "SDK:\n\n    # Object instantiation:\n"
    cobra_str += "{0}".format(cobra_str2)
    cobra_str += "{0}\n".format(qstring)
    return cobra_str


def handle_class(purl, qstring):
    if purl.classnode != "":
        cobra_str = ""
        cobra_str += "    # Cobra does not support APIC based node " + \
                     "queries at this time\n"
    else:
        cobra_str2 = ""
        cobra_str2 += "    # Direct class query:\n"
        cobra_str2 += get_class_query(purl.dn_or_class)
        cobra_str2 += "\n"
        cobra_str = "SDK:\n\n{0}".format(cobra_str2)
        cobra_str += "{0}\n".format(qstring)
    return cobra_str

def handle_aaa_login(purl, qstring):
    # Special case the login, not sure when this would ever be seen though
    cobra_str = "SDK:\n\n    md.login()"
    return cobra_str


def handle_aaa_logout(purl, qstring):
    # Special case the logout.
    cobra_str = "SDK:\n\n    md.logout()"
    return cobra_str


def process_get(url):
    """Process a get request log message."""
    if 'subscriptionRefresh.json' in url or 'aaaRefresh.json' in url:
        return
    purl = apic_rest_urlparse(url)
    qstring = parse_apic_options_string(purl.query)
    supported_api_methods = {
        'mo': handle_mo,
        'class': handle_class,
        'aaaLogin': handle_aaa_login,
        'aaaLogout': handle_aaa_logout,
    }

    try:
        cobra_str = supported_api_methods[purl.api_method](purl, qstring)
    except KeyError:
        cobra_str = "\n# api method {0} is not supported yet".format(
            purl.api_method)

    logging_str = "GET URL: {0}\n{1}".format(url, cobra_str)
    logging.debug(logging_str)


def process_post(url, payload):
    """Process a post request log message."""
    purl = apic_rest_urlparse(url)
    qstring = parse_apic_options_string(purl.query)
    arya_inst = arya()
    cobra_str = arya_inst.getpython(jsonstr=payload, brief=True)
    cobra_str2 = ""
    for line in cobra_str.split("\n"):
        cobra_str2 += "    {0}\n".format(line)
    return cobra_str2

def undefined(**kwargs):
    """Process an undefined logging message."""
    process_get(kwargs['data']['url'])


def GET(**kwargs):   # pylint:disable=invalid-name
    """Process a GET logging message."""
    process_get(kwargs['data']['url'])


def POST(**kwargs):  # pylint:disable=invalid-name
    """Process a POST logging message."""
    cobra_str = process_post(kwargs['data']['url'], kwargs['data']['payload'])
    logging_str = "POST URL: %s\nPOST Payload:\n%s\nSDK:\n\n%s"
    logging.debug(logging_str, kwargs['data']['url'], 
                  kwargs['data']['payload'], cobra_str)


def EventChannelMessage(**kwargs):  # pylint:disable=C0103,W0613
    """Process an event channel logging message."""
    pass


def start_server(args):
    """Start the server threads."""
    # This is used to store the certificate filename
    cert = ""

    # Setup a signal handler to catch control-c and clean up the cert temp file
    # No way to catch sigkill so try not to do that.
    # noinspection PyUnusedLocal
    def sigint_handler(sig, frame):  # pylint:disable=unused-argument
        """A signal handler for interrupt."""
        if not args.cert:
            try:
                os.unlink(cert)
            except OSError:  # pylint:disable=pointless-except
                pass
        print "Exiting..."
        sys.exit(0)

    http_server = None
    https_server = None

    if args.single_server is not None:
        SimpleAciUiLogServer.prettyprint = args.nice_output
        SimpleAciUiLogServer.indent = args.indent
    else:
        ThreadingSimpleAciUiLogServer.prettyprint = args.nice_output
        ThreadingSimpleAciUiLogServer.indent = args.indent

    if args.single_server is None:
        http_server = ThreadingSimpleAciUiLogServer(("", args.port),
                                                log_requests=args.logrequests,
                                                location=args.location,
                                                excludes=args.exclude)
    elif args.single_server == 'http':
        http_server = SimpleAciUiLogServer(("", args.port),
                                           log_requests=args.logrequests,
                                           location=args.location,
                                           excludes=args.exclude)
    if http_server:
        # register our callback functions
        http_server.register_function(POST)
        http_server.register_function(GET)
        http_server.register_function(undefined)
        http_server.register_function(EventChannelMessage)

    if not args.cert and (args.single_server is None or
                          args.single_server == 'https'):
        # Workaround ssl wrap socket not taking a file like object
        cert_file = tempfile.NamedTemporaryFile(delete=False)
        cert_file.write(SERVER_CERT)
        cert_file.close()
        cert = cert_file.name
        print("\n+++WARNING+++ Using an embedded self-signed certificate " +
              "for HTTPS, this is not secure.\n")
    else:
        cert = args.cert

    if args.single_server is None:
        https_server = ThreadingSimpleAciUiLogServer(("", args.sslport),
                                                 cert=cert,
                                                 location=args.location,
                                                 log_requests=args.logrequests,
                                                 excludes=args.exclude,
                                                 request_types=args.request_type)
    elif args.single_server == 'https':
        https_server = SimpleAciUiLogServer(("", args.sslport),
                                            cert=cert,
                                            location=args.location,
                                            log_requests=args.logrequests,
                                            excludes=args.exclude,
                                            request_types=args.request_type)

    if https_server:
        # register our callback functions
        https_server.register_function(POST)
        https_server.register_function(GET)
        https_server.register_function(undefined)
        https_server.register_function(EventChannelMessage)

    signal.signal(signal.SIGINT, sigint_handler)  # Or whatever signal

    # This simply sets up a socket for UDP which has a small trick to it.
    # It won't send any packets out that socket, but this will allow us to
    # easily and quickly interogate the socket to get the source IP address
    # used to connect to this subnet which we can then print out to make for
    # and easy copy/paste in the APIC UI.
    ip_addr = [(s.connect((args.apicip, 80)), s.getsockname()[0], s.close())
               for s in [socket.socket(socket.AF_INET,
                                       socket.SOCK_DGRAM)]][0][1]
    print("serving at:")  # pylint:disable=C0325
    if http_server:
        print("http://{0}:{1}{2}".format(str(ip_addr), str(args.port),
                                         str(args.location)))
    if https_server:
        print("https://{0}:{1}{2}".format(str(ip_addr), str(args.sslport),
                                          str(args.location)))

    print("")  # pylint:disable=C0325
    print("Make sure your APIC(s) are configured to send log messages: " +
          "welcome username -> Start Remote Logging")
    if args.single_server == 'http':
        print("Note: If you connect to the APIC GUI via HTTPS, you need to " +
              "start and use the HTTPS server.")
    elif args.single_server is None:
        print("Note: If you connect to your APIC via HTTPS, configure the " +
              "remote logging to use the https server.")
    print("")

    if http_server and https_server:
        serve_forever([http_server, https_server])
    elif http_server:
        serve_forever([http_server])
    elif https_server:
        serve_forever([https_server])


def main():
    """The main function run when AryaLogger is run in standalone mode."""
    parser = ArgumentParser('Archive APIC Rest API calls in the PythonSDK ' +
                            'syntax')
    parser.add_argument('-a', '--apicip', required=False, default='8.8.8.8',
                        help=('If you have a multihomed system, where the ' +
                              'apic is on a private network, the server ' +
                              'will print the ip address your local system ' +
                              'has a route to 8.8.8.8.  If you want the ' +
                              'server to print a more accurate ip address ' +
                              'for theserver you can tell it the apicip ' +
                              'address.'))

    parser.add_argument('-c', '--cert', type=str, required=False,
                        help=('The server certificate file for ssl ' +
                              'connections, default="server.pem"'))

    parser.add_argument('-e', '--exclude', action='append', nargs='*',
                        default=[], choices=['subscriptionRefresh',
                                             'aaaRefresh', 'aaaLogout',
                                             'HDfabricOverallHealth5min-0',
                                             'topInfo', 'all'],
                        help=('Exclude certain types of common "noise" ' +
                              'queries.'))

    parser.add_argument('-i', '--indent', type=int, default=2, required=False,
                        help=('The number of spaces to indent when pretty ' +
                              'printing'))

    parser.add_argument('-l', '--location', default="/apiinspector",
                        required=False,
                        help=('Location that transaction logs are being ' +
                              'sent to, default=/apiinspector'))

    parser.add_argument('-n', '--nice-output', action='store_true',
                        default=False, required=False,
                        help='Pretty print the response and payload')

    parser.add_argument('-po', '--port', type=int, required=False,
                        default=8987,
                        help='Local port to listen on, default=8987')

    parser.add_argument('-r', '--logrequests', action='store_true',
                        default=False, required=False,
                        help=('Log server requests and response codes to ' +
                              'standard error'))
 
    parser.add_argument('-s', '--sslport', type=int, required=False,
                        default=8443,
                        help=('Local port to listen on for ssl connections, ' +
                              ' default=8443'))

    parser.add_argument('-si', '--single-server', type=str, required=False,
                        default=None, choices=['http', 'https'],
                        help=('Only start either the http server or the ' +
                              'https server, the default is to start both'))

    parser.add_argument('-ty', '--request-type', action='append', nargs='*',
                        default=[], choices=['POST', 'GET', 'undefined',
                                             'EventChannelMessage', 'all'],
                        help=('Only log specific request-types, default is ' +
                              'log all that are supported.'))

    args = parser.parse_args()

    logging.basicConfig(level=logging.DEBUG,
                        format='%(asctime)s - \n%(message)s')

    if args.exclude:
        # Flatten the list
        args.exclude = [val for sublist in args.exclude for val in sublist]
        if 'all' in args.exclude:
            args.exclude = ['subscriptionRefresh', 'aaaRefresh', 'aaaLogout',
                            'HDfabricOverallHealth5min-0', 'topInfo']

    if args.request_type:
        # Flatten the list
        args.request_type = [val for sublist in args.request_type for val in 
                                 sublist]
        if 'all' in args.request_type:
            args.request_type = ['POST', 'GET', 'undefined',
                                 'EventChannelMessage']
    else:
        args.request_type = ['POST', 'GET', 'undefined', 'EventChannelMessage']

    start_server(args)


if __name__ == '__main__':
    main()
